packages/framework/src/client/__tests__/resource-request.spec.js

```diff
@@ -0,0 +1,218 @@
+
+			expect( request.getStatus( now ) ).toBe( STATUS.scheduled );
+			expect( request.getTimeLeft() ).toBeLessThan( 2 * MINUTE );
+		} );
+
+		it( 'returns a negative value for an overdue request', () => {
+			const requirement = { freshness: 2 * MINUTE, timeout: 3 * SECOND };
+			const resourceState = { lastReceived: threeMinutesAgo };
+
+			const request = new ResourceRequest( 'resource1', requirement, resourceState, now );
+
+			expect( request.getStatus( now ) ).toBe( STATUS.overdue );
+			expect( request.getTimeLeft( now ) ).toBe( -( 1 * MINUTE ) );
+		} );
+	} );
+
+	describe( 'requested', () => {
+		it( 'progresses through success flow correctly', () => {
+			const request = new ResourceRequest( 'resource1', {}, {}, tenSecondsAgo );
+
+			expect( request.getStatus( now ) ).toBe( STATUS.overdue );
+
+			let promiseResolve;
+			const promise = new Promise( ( resolve ) => {
+				promiseResolve = resolve;
+			} );
+
+			request.requested( promise );
+			expect( request.getStatus() ).toBe( STATUS.inFlight );
+
+			promiseResolve();
+
+			return promise.then( () => {
+				expect( request.getStatus() ).toBe( STATUS.complete );
+			} );
+		} );
+
+		it( 'progresses through error flow correctly', () => {
+			const request = new ResourceRequest( 'resource1', {}, {}, tenSecondsAgo );
+
+			expect( request.getStatus( now ) ).toBe( STATUS.overdue );
+
+			let promiseReject;
+			const promise = new Promise( ( resolve, reject ) => {
+				promiseReject = reject;
+			} );
+
+			request.requested( promise, now );
+			expect( request.getStatus( now ) ).toBe( STATUS.inFlight );
+
+			promiseReject( 'error message' );
+
+			return promise.catch( () => {
+				expect( request.getStatus( now ) ).toBe( STATUS.failed );
+				expect( request.error ).toBe( 'error message' );
+			} );
+		} );
+	} );
+
+	describe( 'addRequirement', () => {
+		it( 'reschedules earlier', () => {
+			const requirement1 = { freshness: 5 * MINUTE, timeout: 3 * SECOND };
+			const requirement2 = { freshness: 4 * MINUTE, timeout: 3 * SECOND };
+			const resourceState = { lastReceived: threeMinutesAgo };
+
+			const request = new ResourceRequest( 'resource1', requirement1, resourceState, tenSecondsAgo );
+			expect( request.getStatus( now ) ).toBe( STATUS.scheduled );
+			expect( request.getTimeLeft( now ) ).toBe( 2 * MINUTE );
+
+			expect( request.addRequirement( requirement2, resourceState, now ) ).toBeTruthy();
+			expect( request.getStatus( now ) ).toBe( STATUS.scheduled );
+			expect( request.getTimeLeft( now ) ).toBe( 1 * MINUTE );
+		} );
+
+		it( 'does not reschedule if request is already adequate', () => {
+			const requirement1 = { freshness: 4 * MINUTE, timeout: 3 * SECOND };
+			const requirement2 = { freshness: 5 * MINUTE, timeout: 3 * SECOND };
+			const resourceState = { lastReceived: threeMinutesAgo };
+
+			const request = new ResourceRequest( 'resource1', requirement1, resourceState, tenSecondsAgo );
```

Comment:
What does `tenSecondsAgo` mean here? From what I can tell, using anything other than `now` is for testing purposes only, and generally `ResourceRequest` will be initialized with the value `now`, is that correct?

Reply:
When I test timing, I try to follow a convention of "now" being the end of the test, evaluating things that have happened earlier. This is why I calculate timings in the past for test setup. I like that better than having times in the future.

And yes, injecting a "now" time as an optional parameter is a tool I use fairly often in testing. It seems to work well and allows me to measure time accurately without having to delay any tests by any real time.